django/utils/text.py

```diff
@@ -103,38 +110,46 @@ def handle_endtag(self, tag):
             self.tags.remove(tag)
 
     def handle_data(self, data):
-        if self.func == "words":
-            data = re.split(r"(?<=\S)\s(?=\S)", data)
-            output = escape(" ".join(data[: self.num]))
-        else:
-            output = escape("".join(data[: self.num]))
-
+        data, output = self.process(data)
         data_len = len(data)
-        if self.num < data_len:
-            if self.func == "words" and self.truncate and self.truncate[0] == " ":
-                output = output.strip()
+        if self.remaining < data_len:
+            self.remaining = 0
             self.output += Truncator.add_truncation_text(output, self.truncate)
-            self.output += "".join([f"</{tag}>" for tag in self.tags])
-            raise self.TruncationCompleted()
-        self.num -= data_len
+            raise self.TruncationCompleted
+        self.remaining -= data_len
         self.output += output
 
-    def _truncate_html(self, func, text, num, truncate):
-        if func not in ("words", "chars"):
-            raise AttributeError(f"func must be 'words' or 'chars', got '{func}'.")
-        if num <= 0:
-            if func == "words":
-                return ""
-            else:
-                return truncate
-        self.func = func
-        self.num = num
-        self.truncate = truncate
+    def feed(self, data):
         try:
-            self.feed(text)
+            super().feed(data)
         except self.TruncationCompleted:
-            pass
-        return self.output
+            self.output += "".join([f"</{tag}>" for tag in self.tags])
+            self.tags.clear()
+            self.reset()
```

Comment:
It seemed more natural to override `.feed()` and then handle the condensed finalisation.

To keep state consistent I also clear out the tags as would be done by `.handle_endtag()` and dispose of any remaining data by calling `.reset()`.